### PR TITLE
Enhances Search with Website and Category

### LINKED ISSUE #34: Add URL as a field for directory service search rpc
Hello and thank you for your work on this excellent project! 

I am interested in searching the directory service for a beneficiary VASP using their URL and/or name, and then disambiguating results by the VASP's country and their category (e.g. atm, exchange, high-risk exchange, etc).

I was wondering if it would be possible to add `url` and `category` as fields for the [directory service search RPC](https://github.com/trisacrypto/testnet/blob/main/proto/trisads/api/v1alpha1/api.proto#L113-L122) and to update the [`Search` function](https://github.com/trisacrypto/testnet/blob/main/pkg/trisads/trisads.go#L278-L315) to "OR" VASP names and urls together and "AND" country and category?

### PR DESCRIPTION
**Pull request by bbengfort**
Implements search on VASP website and category by adding and improving the search functionality of the Directory Service.

Done:

- Adds website and category indices to the leveldb store 
- Adds URL normalization for website lookups
- Ensures that all IVMS101 names are stored in the names index
- Counties and categories now act as a filter
- Synchronizes and reindexes website and category
- Improved insert and remove indices for more flexibility
- Contains search for container indices

Still TODO:

- [x] Update client with new search parameters (and test)

Fixes #34

### CODE REVIEW

File: pkg/trisads/store/leveldb.go [hunk 1]
```diff
@@ -260,15 +262,47 @@ func (s *ldbStore) Search(query map[string]interface{}) (vasps []*pb.VASP, err e
...
+		log.Debug().Strs("website", websites).Msg("search website query")
+		for _, website := range websites {
+			if id := s.websites[website]; id != "" {
+				records[id] = struct{}{}
+			}
+		}
+	}
+
+	// Filter by country
+	// NOTE: if country is not in the index, no records will be returned
```
Comment (rebeccabilbro) [inline pkg/trisads/store/leveldb.go:277]: This is perfect, just what I needed; thank you!

File: cmd/trisads/main.go [hunk 2]
```diff
@@ -142,11 +144,19 @@ func main() {
...
+					Usage: "one or more websites of VASPs to search for",
 				},
 				cli.StringSliceFlag{
 					Name:  "c, country",
-					Usage: "one or more countries of the VASPs to search for",
+					Usage: "one or more countries to filter on",
+				},
+				cli.StringSliceFlag{
+					Name:  "C, category",
+					Usage: "one or more categories to filter on",
```
Comment (rebeccabilbro) [inline cmd/trisads/main.go:159]: Is this the primary entry point for querying or do you imagine users might use `trisads.go` more as library code? Or both?

**Review by rebeccabilbro — APPROVED** [review]
Thank you so much for the quick response and fix; this seems to be exactly what I need!

File: pkg/trisads/store/leveldb.go [hunk 3]
```diff
@@ -260,15 +262,47 @@ func (s *ldbStore) Search(query map[string]interface{}) (vasps []*pb.VASP, err e
...
+
+	// Filter by category
+	// NOTE: if category is not in the index, no records will be returned
+	categories, ok := parseQuery("category", query, normalize)
+	if ok {
+		for _, category := range categories {
+			for record := range records {
+				if !s.categories.contains(category, record, nil) {
+					// Remove the found VASP since it is not in the country index
+					// NOTE: safe to remove during map iteration: https://stackoverflow.com/questions/23229975/is-it-safe-to-remove-selected-keys-from-map-within-a-range-loop
```
[the remark this answers is not in the document]
Comment (bbengfort) [reply pkg/trisads/store/leveldb.go:299]: Won't catch me again!

File: cmd/trisads/main.go (hunk 2 above)
Comment (bbengfort) [reply cmd/trisads/main.go:159]: I think this is more example and debugging code, most search functionality will utilize the library code in `"github.com/trisacrypto/testnet/pkg/trisads/pb/api/v1alpha1"` (and the enums in `"github.com/trisacrypto/testnet/pkg/trisads/pb/models/v1alpha1"`).

**Pull request merged**